python/Ice/bidir/client/client.py

```diff
@@ -0,0 +1,28 @@
+# Copyright (c) ZeroC, Inc.
+
+from EarlyRiser import AlarmClock, ButtonPressed
+
+class MockAlarmClock(AlarmClock):
+
+    def __init__(self, loop):
+        self._needMoreTime = True
+        self._loop = loop
+        self._stopPressed = loop.create_future()
+
+    async def waitForStopPressed(self):
+        await self._stopPressed
+    
+    def stop(self):
+        if not self._stopPressed.done():
+            self._stopPressed.set_result(None)
+
+    def ring(self, message, current):
+        print(f"Dispatching ring request {{ message = '{message}' }}")
+        if self._needMoreTime:
+            print(f"Returning {ButtonPressed.Snooze} to request more time.")
+            self._needMoreTime = False # we only snooze one time
+            return ButtonPressed.Snooze
+        else:
+            self._loop.call_soon_threadsafe(self.stop)
```

Comment:
As I understand it, we use asyncio for this communicator, so all dispatches are in the asyncio loop/thread.

The reason you call `self._loop.call_soon_threadsafe(self.stop)` and not self.stop is because you want to execute self.stop's continuation after the return? Why does it matter?

Reply:
We don't know if a python dispatch method is sync or async until it returns. Dispatch cannot run in asyncio loop thread by default. If a dispatch method returns a coroutine, then we run it using the communicator's event loop adapter. I guess a way to avoid the `call_soon_threadsafe` call is to convert this dispatch method into an `async def`

Reply:
If we turn this dispatch into an async dispatch, we can use `call_soon` instead of `call_soon_threadsafe`. But either way we need to use this pattern, this is roughly equivalent to the `RunContinuationsAsynchronously` in the C# version. We don't want the stop pressed callback to run synchronously otherwise the dispatch would deadlock.